### PR TITLE
Excel toolkit issue85 startup performance

### LINKED ISSUE #85: Excel_Toolkit: Startup performance
Just dropping here to investigate in the near future. Would be good to improve the performance of opening excel. The poor performance relative to the other UIs is owing to the fact that we need to wrap each method/type/etc. in an Ecxel formula from startup time as opposed to just enumerating all possibilities into a menu, to be dealt with on-demand.

There may be a way we can do it smarter.

### LINKED ISSUE #100: Excel_Toolkit: Crashes on exit
After you close Excel, the 'not responding' dialogue opens, crashing out of Excel. No test file as it's difficult to recreate - doesn't seem to happen with all files/new blank documents. @awakeman as mentioned, checking any COM release methods would be good.

### LINKED ISSUE #101: Excel_Toolkit: Create methods that return system types are filed incorrectly
As of #96 the leading part of the namespace for Create methods uses the namespace of the type it returns, in order to file methods where they are expected. an example being `Create.Structure.Elements.Bar` as opposed to `Create.Structure.Bar`, the create method for which resides in `BH.Engine.Structure.Create` and not `BH.Engine.Structure.Elements.Create`.

As such methods like `BH.Engine.Reflection.Create.Type(string)` (which returns `Type`) ends up as `Create.System.Type`, not `Create.Reflection.Type`.

### PR DESCRIPTION
**Pull request by awakeman**
### Issues addressed by this PR

Closes #85 

~~It now defers adding items to the context menus until a workbook is opened and does so in background threads, I believe there is some locking going on in the COM interface still as it is a little sluggish whilst it does it but it does seem to complete faster than when it was done fully synchronously. Not entirely happy with it for this reason but struggling to find a better method. It was my hope that there would be some event to hook into as each sub-menu was opened so that we could build the menu on demand but it appears the MS Office API doesn't provide one.~~

Removes context menu entirely.

Closes #101 

As of #96 the leading part of the namespace for Create methods uses the namespace of the type it returns, in order to file methods where they are expected. an example being `Create.Structure.Elements.Bar` as opposed to `Create.Structure.Bar`, the create method for which resides in `BH.Engine.Structure.Create` and not `BH.Engine.Structure.Elements.Create`.

As such methods like `BH.Engine.Reflection.Create.Type(string)` (which returns `Type`) ends up as `Create.System.Type`, not `Create.Reflection.Type`.

This PR corrects this.

Closes #100 

Sometimes excel process continues after all windows are closed, caused by there being unreleased COM objects. Added a commit to make sure we release them correctly to avoid this.

### Test files
#85  [profiling.xlsx](https://github.com/BHoM/Excel_Toolkit/files/3271384/profiling.xlsx)

You will need to compare with `Excel_Toolkit-issue85-BaselinePerformanceWithProfiling` branch, which is just mater with the additional commit that adds the profiling stuff `cherry-pick`ed in.

#101 [namespaces.xlsx](https://github.com/BHoM/Excel_Toolkit/files/3277268/namespaces.xlsx)

This has methods from all the `BHoM_Engine` projects, if you calculate on `master` you should get `#NAME!` errors for some of them, this PR should resolve that.

#100 Hard to create a testfile for a transient failure upon close.

### CODE REVIEW

**Comment by rwemay** [comment]
I built this and tried to open Excel/the multistorey spreadsheet in test library and it crashed twice on opening. On the third attempt Excel opened the sheet a bit faster than it was doing, but was then unresponsive for a minute + once it was open. 

I don't think this is the solution - what's the next option? 

If it's the context menu that's the problem, why don't we just disable this for now? With global search (is that in the works?) and in-cell formulae, we'll have to manage until there is a solution.

**Comment by michaldengusiak** [comment]
Could we add this a part of Excel Ribbon?
This would allow us to open excel quickly and then enable via Ribbon?

**Comment by awakeman** [comment]
Yeah, realistically I think the next solution is to disable the menu but give a mechanism for the user to manually enable it. The performance of the COM API (the only way to manipulate the context menus/ribbon) is bad and we're registering thousands of buttons 99% of which are likely unused for any given session. As I said above excel gives us no way to to this lazily in order to not do the redundant work.

Will back this out and do that.

**Comment by awakeman** [comment]
@rwemay, @michaldengusiak I think I have a bit of a middle-of-the road solution.

First off I've reverted to only presenting the BHoM menus on the normal "Cell" context menu, so it will not work from print preview or tables anymore and secondly I don't populate menus with more than 200 leaf nodes in the menu tree automatically, the user is instead presented with a button to load them on demand. 

image

Clicking this does close the menu so you have to navigate it twice to get to the thing you want.

Coupled with bumping up the priority of the Global Search (#56) though should release some of the pain here.

For me this brings the load time down below 30s.

**Comment by awakeman** [comment]
I've added some profiling capability so that we can capture the load times and break it down to some critical parts. See below:

Before:
image

After:
image

Here is the sheet to test for yourself: [profiling.xlsx](https://github.com/BHoM/Excel_Toolkit/files/3271384/profiling.xlsx)

You will need to compare with `Excel_Toolkit-issue85-BaselinePerformanceWithProfiling` branch, which is just master with the additional commit that adds the profiling stuff `cherry-pick`ed in.

**Comment by rwemay** [comment]
Please add 'plugin disabled' to the profiles, as this is obviously the base benchmark.

**Comment by awakeman** [comment]
I can't use the BHoM to time Excel opening without the BHoM, needless to say it is still considerably faster.

**Comment by rwemay** [comment]
Thanks @awakeman , as discussed, I think adding the non-BHoM loading baseline is important. If it cant be coded/calculated exactly, I would start by simply timing the opening from double click, with BHoM disabled. We do need to measure the impact we make to user experience so we can communicate this/test against it. Hopefully its all short term anyway, once we resolve the load performance in a scalable manner.

**Comment by awakeman** [comment]
@rwemay all that is being timed above is the time it takes to load the BHoM Addin, so the reference for when it is disabled is 0. i.e. the above is how much time is being added by the BHoM plugin under the conditions of my PC. The specific number will vary based on a number of factors specific to the user and specific to the situation (e.g. how loaded the CPU is at the moment the application is opened). The important part is the relative difference, 5-6x improvement.

**Comment by rwemay** [comment]
This appears to force the view in open mode to be 'normal view' even if files is saved in sheet/print view(?) Test file below. Load time is still relatively long. Please make sure to benchmark with BHoM not loaded at all - that would rule out any events/lag with metrics not currently measured/that we aren't aware of. 

https://burohappold.sharepoint.com/:f:/r/sites/BHoM/02_Current/12_Scripts/01_Test%20Scripts/Excel_Toolkit/Excel_Toolkit-Issue97-Null%20causing%20lists%20to%20fail?csf=1&e=VPaZ18

**Comment by awakeman** [comment]
Looking into this now. As for expected behaviour it is worth looking at #46, #78 and #79 (and the PRs that solve them) to see discussion on how to treat empty inputs and the conflicting possibilities of how we could do so.

**Comment by awakeman** [comment]
Had a conversation with @rwemay and we agreed to kill the context menu entirely for now. For him the delayed loading wasn't working anyway.

Focus will now move to Global search. Would be good to get some fresh eyes on bringing back the context menu without poor performance in the future @al-fisher @IsakNaslundBh.

**Comment by michaldengusiak** [comment]
@awakeman  `context menu` you mean search using right click ? this means we will be only able to type formulas for time being?...

**Comment by rwemay** [comment]
Thats correct @michaldengusiak . Until there is a working and tested version of the context menu it will be disabled. In interim we will need to use formulae and/or the ctrl shift B menu when it implemented. If you have time to help @awakeman  with either/or solution, feel free to pitch in.

File: Excel_UI/UI/Templates/SelectorMenu_CommandBar.cs [hunk 1]
```diff
@@ -75,8 +75,21 @@ private void AppendMenuTree(Tree<T> tree, CommandBarControls menu)
...
+                            foreach (Tree<T> childTree in tree.Children.Values.OrderBy(x => x.Name))
+                                AppendMenuTree(childTree, treeMenu);
+                            Ctrl.Visible = false;
+                        });
+                    } else
+                    {
+                        foreach (Tree<T> childTree in tree.Children.Values.OrderBy(x => x.Name))
+                            AppendMenuTree(childTree, treeMenu);
+                    }
                 }
```
Comment (rwemay) [inline Excel_UI/UI/Templates/SelectorMenu_CommandBar.cs]: Does this relate to the context menu/trying to limit the number of items loaded, only? If so please remove - it didn't work when I tested, and in any case is not relevant if we're not loading the context menu.

File: Excel_UI/UI/Templates/SingleOptionCallerFormula.cs [hunk 2]
```diff
@@ -42,14 +42,6 @@ public abstract class SingleOptionCallerFormula : CallerFormula
...
-            {
-                var btn = menu.Controls.Add(MsoControlType.msoControlButton, Temporary: true) as CommandBarButton;
-                btn.Tag = Guid.NewGuid().ToString();
-                btn.Caption = Caller.Name;
-                btn.Click += OnClick;
-                m_btns.Add(btn);
-            }
         }
 
         /*******************************************/
```
Comment (rwemay) [inline Excel_UI/UI/Templates/SingleOptionCallerFormula.cs:47]: Should we not remove the method altogether if removing the body?

File: Excel_UI/UI/Templates/CallerFormula.cs [hunk 3]
```diff
@@ -94,35 +93,6 @@ public CallerFormula(FormulaDataAccessor accessor, List<CommandBar> ctxMenus)
...
-                    ) as CommandBarPopup;
-                if (menu == null)
-                {
-                    menu = commandBar.Controls.Add(MsoControlType.msoControlPopup, Temporary: true) as CommandBarPopup;
-                    menu.Caption = "BH." + Category;
-                    menu.Tag = Category;
-                }
-                Menus.Add(menu);
-            }
-
```
Comment (rwemay) [inline Excel_UI/UI/Templates/CallerFormula.cs:121]: Have we exhausted all options of creating the context menu? (e.g. what's the cost of 'FindControl'/is it required? - I assume yes as it looks like the previous single caller did not have this).

**Review by rwemay — CHANGES_REQUESTED** [review]

**Comment by rwemay** [comment]
Looking at the errors and warnings that show up on startup, I'm also wondering if there are gains in resolving/removing some of these (e.g. not trying to load RhinoCommon)? I suspect based on the earlier profiling these aren't big gains. See output file below. 

@awakeman , please make sure all test scripts are saved to the test script folder. I have created one for you for this issue. It is very important to follow this, particularly as test files/scripts may contain a mix of internal/external namespaces, references and information. 

https://burohappold.sharepoint.com/:f:/r/sites/BHoM/02_Current/12_Scripts/01_Test%20Scripts/Excel_Toolkit/Excel_Toolkit-Issue85-Startup%20Performance?csf=1&e=BQo1Bm

File: Excel_UI/UI/Templates/CallerFormula.cs (hunk 3 above)
Comment (awakeman) [reply Excel_UI/UI/Templates/CallerFormula.cs:121]: The single caller inherited it from this one. Yes it is required to create the sub-menu for each category only once.

File: Excel_UI/UI/Templates/SelectorMenu_CommandBar.cs (hunk 1 above)
Comment (awakeman) [reply Excel_UI/UI/Templates/SelectorMenu_CommandBar.cs]: yeah, I will remove.

File: Excel_UI/UI/Templates/SingleOptionCallerFormula.cs (hunk 2 above)
Comment (awakeman) [reply Excel_UI/UI/Templates/SingleOptionCallerFormula.cs:47]: It's a constructor, it still needs to forward its parameters to the base constructor.

**Comment by awakeman** [comment]
Merging in #102 on request of @rwemay 



   
### Issues addressed by this PR


Closes #101

As of #96 the leading part of the namespace for Create methods uses the namespace of the type it returns, in order to file methods where they are expected. an example being `Create.Structure.Elements.Bar` as opposed to `Create.Structure.Bar`, the create method for which resides in `BH.Engine.Structure.Create` and not `BH.Engine.Structure.Elements.Create`.

As such methods like `BH.Engine.Reflection.Create.Type(string)` (which returns `Type`) ends up as `Create.System.Type`, not `Create.Reflection.Type`.

This PR corrects this.




### Test files

[namespaces.xlsx](https://github.com/BHoM/Excel_Toolkit/files/3277268/namespaces.xlsx)

This has methods from all the `BHoM_Engine` projects, if you calculate on `master` you should get `#NAME!` errors for some of them, this PR should resolve that.

### Changelog




### Additional comments

**Comment by awakeman** [comment]
Regarding those errors and warnings i'll raise some issues from them. I believe that the RhinoCommon, Grasshopper and Revit dll's that attempt to load are due to dependencies from the Rhino, Grasshopper and Revit toolkits and are non-fatal.

The errors in registering the Templated `CreateType` and `CreateCustom` formulae would be good to fix but are due to `<`,`>` and `,` being illegal characters in a formula name. As far as I am aware the affected types are very unlikely to be desired in a `FilterQuery`, for example.

Will raise some issues for these.

**Comment by rwemay** [comment]
Great, thanks. Let's add the fixes that we can to this PR as they all relate to start up. Great to get as many of these bugs squashed as possible so we're building on a more solid opening-performance platform.

**Comment by awakeman** [comment]
Closes #100

Sometimes excel process continues after all windows are closed, caused by there being unreleased COM objects. Added a commit to make sure we release them correctly to avoid this.

**Comment by michaldengusiak** [comment]
@awakeman 
I tested this and is building fine and is working. It is faster it takes for me 16 sek compare to 2 min before. Great

The only comment I have because we do not have content menu AND we type in excel differently compate to other UIs is much more difficult to find methods. I hope we find in future some way to load menu or provide search just to make templates. Once is set up loading time should be priority. Maybe we can still have some way to run long version a debug/ script creation option ? 


Excel
image
Dynamo, GH
image
image

**Comment by awakeman** [comment]
Nice catch @michaldengusiak, thanks, that should be fixed now.

**Review by michaldengusiak — APPROVED** [review]
@awakeman 

thanks we have now
`=Create.Environment.Elements.Building?by_String_Double_Double_Double_BuildingType("aa")`

### so normally this is excel rule..
we have method in BHoM that is used in _GH and Dynamo_:
`BH.Engine.Environment.Query.PanelsContainingPoint`

with _Excel_ we delete first two `BH.Engine` or `BH.oM`
`Environment.Query.PanelsContainingPoint`
and change order
`Query.Environment.PanelsContainingPoint`

personally, If I could change  I would prefer to type full 1:1 formula even if is long - would be consistent  but if this is what we agreed then I am  fine...
we just need to make clear documentation to everyone how to type...

**Comment by rwemay** [comment]
@michaldengusiak , I agree with you - 1:1 is better to align across all (current) platforms, *although* in Excel, having the long method/namespace names does make it a bit trickier to nest formulae - the formula quickly gets very long. I would therefore err on the side of shortened. @IsakNaslundBh you've used Excel_Toolkit as much/more than any - what are your thoughts? 

Given that profiling the formula creation has shown that it's not expensive/a governing factor, would it be crazy to keep both....? So you can type with intellisense either BH.Engine.Environment or Environment.Query etc. 

This is a relatively big decision as it will obviously break lots of spreadsheets if we change direction later, but perhaps by allowing both we can take care of this more easily...?

**Review by rwemay — APPROVED** [review]
Approving this so that we can move forwards with further testing, as it addresses one key issue of startup performance (which itself is hindering testing). 

As far as testing - @michaldengusiak  has tested/approved with comments, and I have been using Excel with a couple of workbooks from this branch for a week and it appears to be working.

**Comment by rwemay** [comment]
Merging this and will close, and move the suggestion of supporting multiple namespace formula options as a separate issue for discussion.

**Pull request merged**